File: vibeshield-python/examples/rce_usage.py
"""
VibeShield RCE Pattern Detector — Usage Example

Demonstrates how to detect Remote Code Execution (RCE) vulnerabilities,
sandbox escapes, and dangerous system command calls in Python source code.

Run with: python examples/rce_usage.py
"""

import os
import sys

# Allow importing from src/
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from src.rce_detector import detect_rce_patterns

print("=============================================================")
print("[SECURE] VIBESHIELD RCE SECURITY VALIDATION DEMO")
print("=============================================================\n")

# A set of code payloads to analyze
payloads = {
    "Safe Function": """
def get_user_profile(user_id):
    # Simple dictionary lookup is perfectly safe
    return db.query("SELECT * FROM profiles WHERE id = ?", user_id)
""",
    "Critical: eval() with Dynamic Input": """
def run_calculator(expr):
    # CRITICAL: executing dynamic untrusted mathematical expressions
    return eval("2 + " + expr)
""",
    "High: eval() with Static Literal": """
def test_static():
    # HIGH: eval used with static string
    return eval("1 + 1")
""",
    "Critical: Subprocess shell=True with Dynamic Input": """
def ping_host(host):
    # CRITICAL: shell=True with concatenated variable
    subprocess.run("ping -c 1 " + host, shell=True)
""",
    "High: Subprocess shell=True with Static Input": """
def list_files():
    # HIGH: shell=True with static string
    subprocess.run("ls -la", shell=True)
""",
    "Warning: Subprocess with Dynamic Input (No Shell)": """
def run_command(arg):
    # WARNING: passing dynamic argument list
    subprocess.run(["ls", arg])
""",
    "High: Dynamic getattr() on Variables": """
def invoke_method(obj, method_name):
    # HIGH: dynamic getattr lookup on variable
    func = getattr(obj, method_name)
    return func()
""",
    "Warning: Sandbox Escape (globals/locals)": """
def inspect_context():
    # WARNING: exposing global symbol table
    return globals()
"""
}

for title, code in payloads.items():
    print(f"--- Scenario: {title} ---")
    result = detect_rce_patterns(code)
    print(f"Safe: {result['safe']}")
    if result["findings"]:
        print(f"Findings ({len(result['findings'])}):")
        for finding in result["findings"]:
            print(f"  Line {finding['line']}: [{finding['severity'].upper()}] - {finding['pattern']}")
            print(f"    Original code (masked): {finding['original_code']}")
            print(f"    Suggestion: {finding['suggestion']}")
    else:
        print("  No issues found.")
    print()

File: vibeshield-ts/examples/rce-usage.ts
/**
 * VibeShield RCE Pattern Detector — Usage Example
 *
 * This file demonstrates how to use the RCE Pattern Detector to analyze
 * source code files for potential Remote Code Execution vulnerabilities.
 *
 * Run with: npx tsx examples/rce-usage.ts
 */

import { detectRcePatterns } from '../src/core/rce-detector.js';

console.log('=============================================================');
console.log('🛡️ VIBESHIELD RCE PATTERN DETECTOR DEMO');
console.log('=============================================================\n');

// Helper to log detection results
function runAnalysis(caseName: string, code: string) {
  console.log(`--- ${caseName} ---`);
  console.log('Code to analyze:');
  console.log(code.trim().split('\n').map(l => `  | ${l}`).join('\n'));
  
  const result = detectRcePatterns(code);
  
  console.log(`Safe: ${result.safe ? '✅ Yes' : '❌ No'}`);
  if (result.findings.length > 0) {
    console.log('Findings:');
    result.findings.forEach((finding, index) => {
      console.log(`  [${index + 1}] Line ${finding.line} | Pattern: ${finding.pattern}`);
      console.log(`      Severity: ${finding.severity.toUpperCase()}`);
      console.log(`      Masked Code: ${finding.originalCode}`);
      console.log(`      Suggestion: ${finding.suggestion}`);
    });
  }
  console.log('\n');
}

// Case 1: Safe code
runAnalysis(
  'CASE 1: Safe standard code',
  `
const fs = require("fs");
setTimeout(() => {
  console.log("Safe timer executed.");
}, 1000);
  `
);

// Case 2: Static eval vs Dynamic eval
runAnalysis(
  'CASE 2: Static and Dynamic eval()',
  `
// Static eval (discouraged but not user-controlled input)
eval("const a = 1;");

// Dynamic eval (highly dangerous, critical RCE risk)
const userInput = "process.exit()";
eval("run(" + userInput + ")");
  `
);

// Case 3: Dangerous child_process commands
runAnalysis(
  'CASE 3: Child Process Execution',
  `
const { exec, spawn } = require("child_process");

// Warning: standard spawn
spawn("ls", ["-la"]);

// Critical: command built with string concatenation
exec("rm -rf " + targetDirectory);

// Critical: spawn with shell option enabled
spawn("git", ["clone", repo], { shell: true });
  `
);

// Case 4: Sandbox Escape attempt and comments
runAnalysis(
  'CASE 4: Sandbox Escape & Ignored Comments',
  `
// We should ignore commented-out issues like:
// eval(dangerousInput);

/*
  child_process.exec(userInput);
*/

// Actual escape attempt
const binding = process.binding("fs");
  `
);

console.log('=============================================================');
console.log('💡 TIP: Run this detector pre-commit or in CI pipelines to block');
console.log('   accidental introduction of dynamic code execution patterns.');
console.log('=============================================================');
